jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -394,9 +398,27 @@ public Object run() {
  *
  * @see			java.lang.Class
  */
+@SuppressWarnings("removal")
 @CallerSensitive
-public static Class<?> forName(String className) throws ClassNotFoundException {
-	@SuppressWarnings("removal")
+public static Class<?> forName(String className) throws ClassNotFoundException
+{
+/*[IF JAVA_SPEC_VERSION >= 18]*/
+	return forNameHelper(className, null, false);
+}
+
+@SuppressWarnings("removal")
+@CallerSensitiveAdapter
+private static Class<?> forName(String className, Class<?> caller) throws ClassNotFoundException
```

Comment:
Can the public api call this private method like https://github.com/eclipse-openj9/openj9/blob/5aac7ed169263e9be0ede84a71ff5a388bec393e/jcl/src/java.base/share/classes/java/lang/ClassLoader.java#L1408-L1412
In addition, what's the expected caller of this private method?

Reply:
> Can the public api call this private method like 

No. Calls such as `getStackClass(2)` and `getStackClassLoader(2)` won't work if the private method is called from the public method.

> what's the expected caller of this private method?

MethodHandle code to support JEP416 will use these private methods since it will already know the `caller`. Search for `caller class` in [JEP416 support](https://github.com/ibmruntimes/openj9-openjdk-jdk18/commit/ed8f09d38f1f29b69a17a3778325143d40d76d11) ... there are instances of `mh.invoke` with `caller` as the last parameter.

Reply:
Can following (roughly) code snippet work?
```
@CallerSensitive
public static Class<?> forName(String className) throws ClassNotFoundException
{
/*[IF JAVA_SPEC_VERSION >= 18]*/
	return forName(className, getStackClass(1));
}

@CallerSensitiveAdapter
private static Class<?> forName(String className, Class<?> caller) throws ClassNotFoundException
{
/*[ENDIF] JAVA_SPEC_VERSION >= 18 */
	@SuppressWarnings("removal")
	SecurityManager sm = null;
	/**
	 * Get the SecurityManager from System.  If the VM has not yet completed bootstrapping (i.e., J9VMInternals.initialized is still false)
	 * sm is kept as null without referencing System in order to avoid loading System earlier than necessary.
	 */
	if (J9VMInternals.initialized) {
		sm = System.getSecurityManager();
	}
	ClassLoader callerClassLoader = null;
	if (null == sm) {
/*[IF JAVA_SPEC_VERSION >= 18]*/
		return forNameImpl(className, true, caller.internalGetClassLoader());
/*[ELSE] JAVA_SPEC_VERSION >= 18
		return forNameImpl(className, true, ClassLoader.callerClassLoader());
/*[ENDIF] JAVA_SPEC_VERSION >= 18 */
	}
/*[IF JAVA_SPEC_VERSION <= 17]*/
	Class<?> caller = getStackClass(1);
/*[ENDIF] JAVA_SPEC_VERSION <= 17 */
	if (null != caller) {
		callerClassLoader = caller.getClassLoaderImpl();
	}
	Class<?> c = forNameImpl(className, false, callerClassLoader);
	forNameAccessCheck(sm, caller, c);
	J9VMInternals.initialize(c);
	return c;
}
```

Reply:
**Added another commit** which reflects the above suggested change: `Reuse private Class.forName(..., caller) from public Class.forName(...)`. Running a personal build to check for functional issues.

With this commit, `getStackClass(1)` will always be called. This will cause a perf regression because `getStackClass(1)` is not needed if `(sm == null)`.

Reply:
> With this commit, getStackClass(1) will always be invoked in the three Class.forName methods. This will cause a perf regression because getStackClass(1) is not needed if (sm == null).

In case of `sm == null`, `ClassLoader.callerClassLoader()` invokes `native ClassLoader getStackClassLoader(int depth)` which is a similar native as `getStackClass(1)`, I don't see a perf regression because of this. 
https://github.com/eclipse-openj9/openj9/blob/2a80e4c17a31f8e40cee409753cbfe2e8f2a9bb5/jcl/src/java.base/share/classes/java/lang/Class.java#L408-L410

Reply:
There are three variants of `forName`:
1. `forName(String className)`
2. `forName(Module module, String name)`
3. `forName(String className, boolean initializeBoolean, ClassLoader classLoader)`

Only the first variant invokes the native `getStackClassLoader(int depth)`. The remaining two variants do not rely on a native method to get the `ClassLoader`. The perf regression will happen for the last two variants. The suggested change has been applied only for the first variant.